From: Stained Glass OS <ke7oxh@gmail.com>
Subject: [PATCH] msvcp90: Add the older numpunct, time_get::_Getcat and basic_ostream forms of msvcp60 to msvcp80.

The older export forms msvcp60 .. msvcp80 have where the msvcp90
sources' functions take one more argument: the numpunct<char/wchar_t/
unsigned short> constructors from a _Locinfo or a locale name and _Init
without the "use the default names" flag (it is off), time_get::_Getcat
without a locale (the classic one), and msvcp70 .. msvcp80's
basic_ostream<wchar_t>(_Uninitialized) without the flag (the standard
streams' constructor: always added to the standard streams). Also maps
ARM names whose i386 twins earlier patches implemented.
---
 dlls/msvcp60/msvcp60.spec | 16 ++++-----
 dlls/msvcp70/msvcp70.spec | 92 +++++++++++++++++++++++------------------------
 dlls/msvcp71/msvcp71.spec | 56 ++++++++++++++---------------
 dlls/msvcp80/msvcp80.spec | 74 +++++++++++++++++++-------------------
 dlls/msvcp90/ios.c        | 12 +++++++
 dlls/msvcp90/locale.c     | 91 ++++++++++++++++++++++++++++++++++++++++++++++
 6 files changed, 222 insertions(+), 119 deletions(-)

diff --git a/dlls/msvcp60/msvcp60.spec b/dlls/msvcp60/msvcp60.spec
index b800a10..7e8f6bf 100644
--- a/dlls/msvcp60/msvcp60.spec
+++ b/dlls/msvcp60/msvcp60.spec
@@ -942,14 +942,14 @@
 @ thiscall -arch=i386 ??0?$num_put@GV?$ostreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QAE@I@Z(ptr long) num_put_short_ctor_refs
 @ cdecl -arch=win64 ??0?$num_put@GV?$ostreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QEAA@_K@Z(ptr long) num_put_short_ctor_refs
 @ stub -arch=arm ??0?$numpunct@D@std@@QAA@ABV_Locinfo@1@I@Z
-@ stub -arch=i386 ??0?$numpunct@D@std@@QAE@ABV_Locinfo@1@I@Z
-@ stub -arch=win64 ??0?$numpunct@D@std@@QEAA@AEBV_Locinfo@1@_K@Z
+@ thiscall -arch=i386 ??0?$numpunct@D@std@@QAE@ABV_Locinfo@1@I@Z(ptr ptr long) numpunct_char_ctor_locinfo_old
+@ cdecl -arch=win64 ??0?$numpunct@D@std@@QEAA@AEBV_Locinfo@1@_K@Z(ptr ptr long) numpunct_char_ctor_locinfo_old
 @ cdecl -arch=arm ??0?$numpunct@D@std@@QAA@I@Z(ptr long) numpunct_char_ctor_refs
 @ thiscall -arch=i386 ??0?$numpunct@D@std@@QAE@I@Z(ptr long) numpunct_char_ctor_refs
 @ cdecl -arch=win64 ??0?$numpunct@D@std@@QEAA@_K@Z(ptr long) numpunct_char_ctor_refs
 @ stub -arch=arm ??0?$numpunct@G@std@@QAA@ABV_Locinfo@1@I@Z
-@ stub -arch=i386 ??0?$numpunct@G@std@@QAE@ABV_Locinfo@1@I@Z
-@ stub -arch=win64 ??0?$numpunct@G@std@@QEAA@AEBV_Locinfo@1@_K@Z
+@ thiscall -arch=i386 ??0?$numpunct@G@std@@QAE@ABV_Locinfo@1@I@Z(ptr ptr long) numpunct_short_ctor_locinfo_old
+@ cdecl -arch=win64 ??0?$numpunct@G@std@@QEAA@AEBV_Locinfo@1@_K@Z(ptr ptr long) numpunct_short_ctor_locinfo_old
 @ cdecl -arch=arm ??0?$numpunct@G@std@@QAA@I@Z(ptr long) numpunct_short_ctor_refs
 @ thiscall -arch=i386 ??0?$numpunct@G@std@@QAE@I@Z(ptr long) numpunct_short_ctor_refs
 @ cdecl -arch=win64 ??0?$numpunct@G@std@@QEAA@_K@Z(ptr long) numpunct_short_ctor_refs
@@ -2883,11 +2883,11 @@
 @ thiscall -arch=i386 ?_Init@?$num_put@GV?$ostreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@IAEXABV_Locinfo@2@@Z(ptr ptr) num_put_wchar__Init
 @ cdecl -arch=win64 ?_Init@?$num_put@GV?$ostreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@IEAAXAEBV_Locinfo@2@@Z(ptr ptr) num_put_wchar__Init
 @ stub -arch=arm ?_Init@?$numpunct@D@std@@IAAXABV_Locinfo@2@@Z
-@ stub -arch=i386 ?_Init@?$numpunct@D@std@@IAEXABV_Locinfo@2@@Z
-@ stub -arch=win64 ?_Init@?$numpunct@D@std@@IEAAXAEBV_Locinfo@2@@Z
+@ thiscall -arch=i386 ?_Init@?$numpunct@D@std@@IAEXABV_Locinfo@2@@Z(ptr ptr) numpunct_char__Init_old
+@ cdecl -arch=win64 ?_Init@?$numpunct@D@std@@IEAAXAEBV_Locinfo@2@@Z(ptr ptr) numpunct_char__Init_old
 @ stub -arch=arm ?_Init@?$numpunct@G@std@@IAAXABV_Locinfo@2@@Z
-@ stub -arch=i386 ?_Init@?$numpunct@G@std@@IAEXABV_Locinfo@2@@Z
-@ stub -arch=win64 ?_Init@?$numpunct@G@std@@IEAAXAEBV_Locinfo@2@@Z
+@ thiscall -arch=i386 ?_Init@?$numpunct@G@std@@IAEXABV_Locinfo@2@@Z(ptr ptr) numpunct_wchar__Init_old
+@ cdecl -arch=win64 ?_Init@?$numpunct@G@std@@IEAAXAEBV_Locinfo@2@@Z(ptr ptr) numpunct_wchar__Init_old
 @ cdecl -arch=arm ?_Init@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@IAAXABV_Locinfo@2@@Z(ptr ptr) time_get_char__Init
 @ thiscall -arch=i386 ?_Init@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@IAEXABV_Locinfo@2@@Z(ptr ptr) time_get_char__Init
 @ cdecl -arch=win64 ?_Init@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@IEAAXAEBV_Locinfo@2@@Z(ptr ptr) time_get_char__Init
diff --git a/dlls/msvcp70/msvcp70.spec b/dlls/msvcp70/msvcp70.spec
index 6fdfe58..59554f1 100644
--- a/dlls/msvcp70/msvcp70.spec
+++ b/dlls/msvcp70/msvcp70.spec
@@ -1,10 +1,10 @@
-@ stub -arch=arm ??0?$_Complex_base@M@std@@QAA@ABM0@Z
+@ cdecl -arch=arm ??0?$_Complex_base@M@std@@QAA@ABM0@Z(ptr ptr ptr) complex_float_ctor
 @ thiscall -arch=i386 ??0?$_Complex_base@M@std@@QAE@ABM0@Z(ptr ptr ptr) complex_float_ctor
 @ cdecl -arch=win64 ??0?$_Complex_base@M@std@@QEAA@AEBM0@Z(ptr ptr ptr) complex_float_ctor
-@ stub -arch=arm ??0?$_Complex_base@N@std@@QAA@ABN0@Z
+@ cdecl -arch=arm ??0?$_Complex_base@N@std@@QAA@ABN0@Z(ptr ptr ptr) complex_double_ctor
 @ thiscall -arch=i386 ??0?$_Complex_base@N@std@@QAE@ABN0@Z(ptr ptr ptr) complex_double_ctor
 @ cdecl -arch=win64 ??0?$_Complex_base@N@std@@QEAA@AEBN0@Z(ptr ptr ptr) complex_double_ctor
-@ stub -arch=arm ??0?$_Complex_base@O@std@@QAA@ABO0@Z
+@ cdecl -arch=arm ??0?$_Complex_base@O@std@@QAA@ABO0@Z(ptr ptr ptr) complex_double_ctor
 @ thiscall -arch=i386 ??0?$_Complex_base@O@std@@QAE@ABO0@Z(ptr ptr ptr) complex_double_ctor
 @ cdecl -arch=win64 ??0?$_Complex_base@O@std@@QEAA@AEBO0@Z(ptr ptr ptr) complex_double_ctor
 @ cdecl -arch=arm ??0?$_Mpunct@D@std@@QAA@ABV_Locinfo@1@I_N@Z(ptr long long) mpunct_char_ctor_refs
@@ -197,9 +197,9 @@
 @ cdecl -arch=arm ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAA@PAV?$basic_streambuf@_WU?$char_traits@_W@std@@@1@_N@Z(ptr ptr long long) basic_ostream_wchar_ctor
 @ thiscall -arch=i386 ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAE@PAV?$basic_streambuf@_WU?$char_traits@_W@std@@@1@_N@Z(ptr ptr long long) basic_ostream_wchar_ctor
 @ cdecl -arch=win64 ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QEAA@PEAV?$basic_streambuf@_WU?$char_traits@_W@std@@@1@_N@Z(ptr ptr long long) basic_ostream_wchar_ctor
-@ stub -arch=arm ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAA@W4_Uninitialized@1@@Z
-@ stub -arch=i386 ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAE@W4_Uninitialized@1@@Z
-@ stub -arch=win64 ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QEAA@W4_Uninitialized@1@@Z
+@ cdecl -arch=arm ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAA@W4_Uninitialized@1@@Z(ptr long long) basic_ostream_wchar_ctor_uninitialized_old
+@ thiscall -arch=i386 ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAE@W4_Uninitialized@1@@Z(ptr long long) basic_ostream_wchar_ctor_uninitialized_old
+@ cdecl -arch=win64 ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QEAA@W4_Uninitialized@1@@Z(ptr long long) basic_ostream_wchar_ctor_uninitialized_old
 @ cdecl -arch=arm ??0?$basic_ostringstream@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@QAA@ABV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@H@Z(ptr ptr long long) basic_ostringstream_char_ctor_str
 @ thiscall -arch=i386 ??0?$basic_ostringstream@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@QAE@ABV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@H@Z(ptr ptr long long) basic_ostringstream_char_ctor_str
 @ cdecl -arch=win64 ??0?$basic_ostringstream@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@QEAA@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@H@Z(ptr ptr long long) basic_ostringstream_char_ctor_str
@@ -591,21 +591,21 @@
 @ cdecl -arch=arm ??0?$num_put@_WV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@QAA@I@Z(ptr long) num_put_wchar_ctor_refs
 @ thiscall -arch=i386 ??0?$num_put@_WV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@QAE@I@Z(ptr long) num_put_wchar_ctor_refs
 @ cdecl -arch=win64 ??0?$num_put@_WV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@QEAA@_K@Z(ptr long) num_put_wchar_ctor_refs
-@ stub -arch=arm ??0?$numpunct@D@std@@QAA@ABV_Locinfo@1@I@Z
-@ stub -arch=i386 ??0?$numpunct@D@std@@QAE@ABV_Locinfo@1@I@Z
-@ stub -arch=win64 ??0?$numpunct@D@std@@QEAA@AEBV_Locinfo@1@_K@Z
+@ cdecl -arch=arm ??0?$numpunct@D@std@@QAA@ABV_Locinfo@1@I@Z(ptr ptr long) numpunct_char_ctor_locinfo_old
+@ thiscall -arch=i386 ??0?$numpunct@D@std@@QAE@ABV_Locinfo@1@I@Z(ptr ptr long) numpunct_char_ctor_locinfo_old
+@ cdecl -arch=win64 ??0?$numpunct@D@std@@QEAA@AEBV_Locinfo@1@_K@Z(ptr ptr long) numpunct_char_ctor_locinfo_old
 @ cdecl -arch=arm ??0?$numpunct@D@std@@QAA@I@Z(ptr long) numpunct_char_ctor_refs
 @ thiscall -arch=i386 ??0?$numpunct@D@std@@QAE@I@Z(ptr long) numpunct_char_ctor_refs
 @ cdecl -arch=win64 ??0?$numpunct@D@std@@QEAA@_K@Z(ptr long) numpunct_char_ctor_refs
-@ stub -arch=arm ??0?$numpunct@G@std@@QAA@ABV_Locinfo@1@I@Z
-@ stub -arch=i386 ??0?$numpunct@G@std@@QAE@ABV_Locinfo@1@I@Z
-@ stub -arch=win64 ??0?$numpunct@G@std@@QEAA@AEBV_Locinfo@1@_K@Z
+@ cdecl -arch=arm ??0?$numpunct@G@std@@QAA@ABV_Locinfo@1@I@Z(ptr ptr long) numpunct_short_ctor_locinfo_old
+@ thiscall -arch=i386 ??0?$numpunct@G@std@@QAE@ABV_Locinfo@1@I@Z(ptr ptr long) numpunct_short_ctor_locinfo_old
+@ cdecl -arch=win64 ??0?$numpunct@G@std@@QEAA@AEBV_Locinfo@1@_K@Z(ptr ptr long) numpunct_short_ctor_locinfo_old
 @ cdecl -arch=arm ??0?$numpunct@G@std@@QAA@I@Z(ptr long) numpunct_short_ctor_refs
 @ thiscall -arch=i386 ??0?$numpunct@G@std@@QAE@I@Z(ptr long) numpunct_short_ctor_refs
 @ cdecl -arch=win64 ??0?$numpunct@G@std@@QEAA@_K@Z(ptr long) numpunct_short_ctor_refs
-@ stub -arch=arm ??0?$numpunct@_W@std@@QAA@ABV_Locinfo@1@I@Z
-@ stub -arch=i386 ??0?$numpunct@_W@std@@QAE@ABV_Locinfo@1@I@Z
-@ stub -arch=win64 ??0?$numpunct@_W@std@@QEAA@AEBV_Locinfo@1@_K@Z
+@ cdecl -arch=arm ??0?$numpunct@_W@std@@QAA@ABV_Locinfo@1@I@Z(ptr ptr long) numpunct_wchar_ctor_locinfo_old
+@ thiscall -arch=i386 ??0?$numpunct@_W@std@@QAE@ABV_Locinfo@1@I@Z(ptr ptr long) numpunct_wchar_ctor_locinfo_old
+@ cdecl -arch=win64 ??0?$numpunct@_W@std@@QEAA@AEBV_Locinfo@1@_K@Z(ptr ptr long) numpunct_wchar_ctor_locinfo_old
 @ cdecl -arch=arm ??0?$numpunct@_W@std@@QAA@I@Z(ptr long) numpunct_wchar_ctor_refs
 @ thiscall -arch=i386 ??0?$numpunct@_W@std@@QAE@I@Z(ptr long) numpunct_wchar_ctor_refs
 @ cdecl -arch=win64 ??0?$numpunct@_W@std@@QEAA@_K@Z(ptr long) numpunct_wchar_ctor_refs
@@ -1041,13 +1041,13 @@
 @ cdecl -arch=arm ??1time_base@std@@UAA@XZ(ptr) time_base_dtor
 @ thiscall -arch=i386 ??1time_base@std@@UAE@XZ(ptr) time_base_dtor
 @ cdecl -arch=win64 ??1time_base@std@@UEAA@XZ(ptr) time_base_dtor
-@ stub -arch=arm ??4?$_Complex_base@M@std@@QAAAAV01@ABV01@@Z
+@ cdecl -arch=arm ??4?$_Complex_base@M@std@@QAAAAV01@ABV01@@Z(ptr ptr) complex_float_assign
 @ thiscall -arch=i386 ??4?$_Complex_base@M@std@@QAEAAV01@ABV01@@Z(ptr ptr) complex_float_assign
 @ cdecl -arch=win64 ??4?$_Complex_base@M@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) complex_float_assign
-@ stub -arch=arm ??4?$_Complex_base@N@std@@QAAAAV01@ABV01@@Z
+@ cdecl -arch=arm ??4?$_Complex_base@N@std@@QAAAAV01@ABV01@@Z(ptr ptr) complex_double_assign
 @ thiscall -arch=i386 ??4?$_Complex_base@N@std@@QAEAAV01@ABV01@@Z(ptr ptr) complex_double_assign
 @ cdecl -arch=win64 ??4?$_Complex_base@N@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) complex_double_assign
-@ stub -arch=arm ??4?$_Complex_base@O@std@@QAAAAV01@ABV01@@Z
+@ cdecl -arch=arm ??4?$_Complex_base@O@std@@QAAAAV01@ABV01@@Z(ptr ptr) complex_double_assign
 @ thiscall -arch=i386 ??4?$_Complex_base@O@std@@QAEAAV01@ABV01@@Z(ptr ptr) complex_double_assign
 @ cdecl -arch=win64 ??4?$_Complex_base@O@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) complex_double_assign
 @ cdecl -arch=arm ??4?$_Ctraits@M@std@@QAAAAV01@ABV01@@Z(ptr ptr) std_Ctraits_op_assign
@@ -1263,7 +1263,7 @@
 @ cdecl -arch=arm ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAAAAV01@AAF@Z(ptr ptr) basic_istream_short_read_short
 @ thiscall -arch=i386 ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAEAAV01@AAF@Z(ptr ptr) basic_istream_short_read_short
 @ cdecl -arch=win64 ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QEAAAEAV01@AEAF@Z(ptr ptr) basic_istream_short_read_short
-@ stub -arch=arm ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAAAAV01@AAG@Z
+@ cdecl -arch=arm ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAAAAV01@AAG@Z(ptr ptr) basic_istream_wchar_read_ushort
 @ thiscall -arch=i386 ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAEAAV01@AAG@Z(ptr ptr) basic_istream_wchar_read_ushort
 @ cdecl -arch=win64 ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QEAAAEAV01@AEAG@Z(ptr ptr) basic_istream_wchar_read_ushort
 @ cdecl -arch=arm ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAAAAV01@AAH@Z(ptr ptr) basic_istream_short_read_int
@@ -2513,12 +2513,12 @@
 @ cdecl -arch=win64 ?_Getcat@?$numpunct@G@std@@SA_KPEAPEBVfacet@locale@2@@Z(ptr) numpunct_short__Getcat_old
 @ cdecl -arch=win32 ?_Getcat@?$numpunct@_W@std@@SAIPAPBVfacet@locale@2@@Z(ptr) numpunct_wchar__Getcat_old
 @ cdecl -arch=win64 ?_Getcat@?$numpunct@_W@std@@SA_KPEAPEBVfacet@locale@2@@Z(ptr) numpunct_wchar__Getcat_old
-@ stub -arch=win32 ?_Getcat@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@SAIPAPBVfacet@locale@2@@Z
-@ stub -arch=win64 ?_Getcat@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@SA_KPEAPEBVfacet@locale@2@@Z
-@ stub -arch=win32 ?_Getcat@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@SAIPAPBVfacet@locale@2@@Z
-@ stub -arch=win64 ?_Getcat@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@SA_KPEAPEBVfacet@locale@2@@Z
-@ stub -arch=win32 ?_Getcat@?$time_get@_WV?$istreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@SAIPAPBVfacet@locale@2@@Z
-@ stub -arch=win64 ?_Getcat@?$time_get@_WV?$istreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@SA_KPEAPEBVfacet@locale@2@@Z
+@ cdecl -arch=win32 ?_Getcat@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@SAIPAPBVfacet@locale@2@@Z(ptr) time_get_char__Getcat_old
+@ cdecl -arch=win64 ?_Getcat@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@SA_KPEAPEBVfacet@locale@2@@Z(ptr) time_get_char__Getcat_old
+@ cdecl -arch=win32 ?_Getcat@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@SAIPAPBVfacet@locale@2@@Z(ptr) time_get_short__Getcat_old
+@ cdecl -arch=win64 ?_Getcat@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@SA_KPEAPEBVfacet@locale@2@@Z(ptr) time_get_short__Getcat_old
+@ cdecl -arch=win32 ?_Getcat@?$time_get@_WV?$istreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@SAIPAPBVfacet@locale@2@@Z(ptr) time_get_wchar__Getcat_old
+@ cdecl -arch=win64 ?_Getcat@?$time_get@_WV?$istreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@SA_KPEAPEBVfacet@locale@2@@Z(ptr) time_get_wchar__Getcat_old
 @ cdecl -arch=win32 ?_Getcat@?$time_put@DV?$ostreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@SAIPAPBVfacet@locale@2@@Z(ptr) time_put_char__Getcat_old
 @ cdecl -arch=win64 ?_Getcat@?$time_put@DV?$ostreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@SA_KPEAPEBVfacet@locale@2@@Z(ptr) time_put_char__Getcat_old
 @ cdecl -arch=win32 ?_Getcat@?$time_put@GV?$ostreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@SAIPAPBVfacet@locale@2@@Z(ptr) time_put_short__Getcat_old
@@ -2753,15 +2753,15 @@
 @ cdecl -arch=arm ?_Init@?$num_put@_WV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@IAAXABV_Locinfo@2@@Z(ptr ptr) num_put_wchar__Init
 @ thiscall -arch=i386 ?_Init@?$num_put@_WV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@IAEXABV_Locinfo@2@@Z(ptr ptr) num_put_wchar__Init
 @ cdecl -arch=win64 ?_Init@?$num_put@_WV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@IEAAXAEBV_Locinfo@2@@Z(ptr ptr) num_put_wchar__Init
-@ stub -arch=arm ?_Init@?$numpunct@D@std@@IAAXABV_Locinfo@2@@Z
-@ stub -arch=i386 ?_Init@?$numpunct@D@std@@IAEXABV_Locinfo@2@@Z
-@ stub -arch=win64 ?_Init@?$numpunct@D@std@@IEAAXAEBV_Locinfo@2@@Z
-@ stub -arch=arm ?_Init@?$numpunct@G@std@@IAAXABV_Locinfo@2@@Z
-@ stub -arch=i386 ?_Init@?$numpunct@G@std@@IAEXABV_Locinfo@2@@Z
-@ stub -arch=win64 ?_Init@?$numpunct@G@std@@IEAAXAEBV_Locinfo@2@@Z
-@ stub -arch=arm ?_Init@?$numpunct@_W@std@@IAAXABV_Locinfo@2@@Z
-@ stub -arch=i386 ?_Init@?$numpunct@_W@std@@IAEXABV_Locinfo@2@@Z
-@ stub -arch=win64 ?_Init@?$numpunct@_W@std@@IEAAXAEBV_Locinfo@2@@Z
+@ cdecl -arch=arm ?_Init@?$numpunct@D@std@@IAAXABV_Locinfo@2@@Z(ptr ptr) numpunct_char__Init_old
+@ thiscall -arch=i386 ?_Init@?$numpunct@D@std@@IAEXABV_Locinfo@2@@Z(ptr ptr) numpunct_char__Init_old
+@ cdecl -arch=win64 ?_Init@?$numpunct@D@std@@IEAAXAEBV_Locinfo@2@@Z(ptr ptr) numpunct_char__Init_old
+@ cdecl -arch=arm ?_Init@?$numpunct@G@std@@IAAXABV_Locinfo@2@@Z(ptr ptr) numpunct_wchar__Init_old
+@ thiscall -arch=i386 ?_Init@?$numpunct@G@std@@IAEXABV_Locinfo@2@@Z(ptr ptr) numpunct_wchar__Init_old
+@ cdecl -arch=win64 ?_Init@?$numpunct@G@std@@IEAAXAEBV_Locinfo@2@@Z(ptr ptr) numpunct_wchar__Init_old
+@ cdecl -arch=arm ?_Init@?$numpunct@_W@std@@IAAXABV_Locinfo@2@@Z(ptr ptr) numpunct_wchar__Init_old
+@ thiscall -arch=i386 ?_Init@?$numpunct@_W@std@@IAEXABV_Locinfo@2@@Z(ptr ptr) numpunct_wchar__Init_old
+@ cdecl -arch=win64 ?_Init@?$numpunct@_W@std@@IEAAXAEBV_Locinfo@2@@Z(ptr ptr) numpunct_wchar__Init_old
 @ cdecl -arch=arm ?_Init@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@IAAXABV_Locinfo@2@@Z(ptr ptr) time_get_char__Init
 @ thiscall -arch=i386 ?_Init@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@IAEXABV_Locinfo@2@@Z(ptr ptr) time_get_char__Init
 @ cdecl -arch=win64 ?_Init@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@IEAAXAEBV_Locinfo@2@@Z(ptr ptr) time_get_char__Init
@@ -4920,22 +4920,22 @@
 @ cdecl -arch=win64 ?ignore@?$basic_istream@GU?$char_traits@G@std@@@std@@QEAAAEAV12@_JG@Z(ptr long long) basic_istream_wchar_ignore
 @ cdecl -arch=arm ?ignore@?$basic_istream@_WU?$char_traits@_W@std@@@std@@QAAAAV12@HG@Z(ptr long long) basic_istream_wchar_ignore
 @ thiscall -arch=i386 ?ignore@?$basic_istream@_WU?$char_traits@_W@std@@@std@@QAEAAV12@HG@Z(ptr long long) basic_istream_wchar_ignore
-@ stub -arch=arm ?imag@?$_Complex_base@M@std@@QAAMABM@Z
+@ cdecl -arch=arm ?imag@?$_Complex_base@M@std@@QAAMABM@Z(ptr ptr) complex_float_imag_set
 @ thiscall -arch=i386 ?imag@?$_Complex_base@M@std@@QAEMABM@Z(ptr ptr) complex_float_imag_set
 @ cdecl -arch=win64 ?imag@?$_Complex_base@M@std@@QEAAMAEBM@Z(ptr ptr) complex_float_imag_set
-@ stub -arch=arm ?imag@?$_Complex_base@M@std@@QBAMXZ
+@ cdecl -arch=arm ?imag@?$_Complex_base@M@std@@QBAMXZ(ptr) complex_float_imag_get
 @ thiscall -arch=i386 ?imag@?$_Complex_base@M@std@@QBEMXZ(ptr) complex_float_imag_get
 @ cdecl -arch=win64 ?imag@?$_Complex_base@M@std@@QEBAMXZ(ptr) complex_float_imag_get
-@ stub -arch=arm ?imag@?$_Complex_base@N@std@@QAANABN@Z
+@ cdecl -arch=arm ?imag@?$_Complex_base@N@std@@QAANABN@Z(ptr ptr) complex_double_imag_set
 @ thiscall -arch=i386 ?imag@?$_Complex_base@N@std@@QAENABN@Z(ptr ptr) complex_double_imag_set
 @ cdecl -arch=win64 ?imag@?$_Complex_base@N@std@@QEAANAEBN@Z(ptr ptr) complex_double_imag_set
-@ stub -arch=arm ?imag@?$_Complex_base@N@std@@QBANXZ
+@ cdecl -arch=arm ?imag@?$_Complex_base@N@std@@QBANXZ(ptr) complex_double_imag_get
 @ thiscall -arch=i386 ?imag@?$_Complex_base@N@std@@QBENXZ(ptr) complex_double_imag_get
 @ cdecl -arch=win64 ?imag@?$_Complex_base@N@std@@QEBANXZ(ptr) complex_double_imag_get
-@ stub -arch=arm ?imag@?$_Complex_base@O@std@@QAAOABO@Z
+@ cdecl -arch=arm ?imag@?$_Complex_base@O@std@@QAAOABO@Z(ptr ptr) complex_double_imag_set
 @ thiscall -arch=i386 ?imag@?$_Complex_base@O@std@@QAEOABO@Z(ptr ptr) complex_double_imag_set
 @ cdecl -arch=win64 ?imag@?$_Complex_base@O@std@@QEAAOAEBO@Z(ptr ptr) complex_double_imag_set
-@ stub -arch=arm ?imag@?$_Complex_base@O@std@@QBAOXZ
+@ cdecl -arch=arm ?imag@?$_Complex_base@O@std@@QBAOXZ(ptr) complex_double_imag_get
 @ thiscall -arch=i386 ?imag@?$_Complex_base@O@std@@QBEOXZ(ptr) complex_double_imag_get
 @ cdecl -arch=win64 ?imag@?$_Complex_base@O@std@@QEBAOXZ(ptr) complex_double_imag_get
 @ cdecl -arch=win32 ?imag@std@@YAMABV?$complex@M@1@@Z(ptr) complex_float_imag
@@ -5973,22 +5973,22 @@
 @ cdecl -arch=arm ?readsome@?$basic_istream@_WU?$char_traits@_W@std@@@std@@QAAHPA_WH@Z(ptr ptr long) basic_istream_wchar_readsome
 @ thiscall -arch=i386 ?readsome@?$basic_istream@_WU?$char_traits@_W@std@@@std@@QAEHPA_WH@Z(ptr ptr long) basic_istream_wchar_readsome
 @ cdecl -arch=win64 ?readsome@?$basic_istream@_WU?$char_traits@_W@std@@@std@@QEAA_JPEA_W_J@Z(ptr ptr long) basic_istream_wchar_readsome
-@ stub -arch=arm ?real@?$_Complex_base@M@std@@QAAMABM@Z
+@ cdecl -arch=arm ?real@?$_Complex_base@M@std@@QAAMABM@Z(ptr ptr) complex_float_real_set
 @ thiscall -arch=i386 ?real@?$_Complex_base@M@std@@QAEMABM@Z(ptr ptr) complex_float_real_set
 @ cdecl -arch=win64 ?real@?$_Complex_base@M@std@@QEAAMAEBM@Z(ptr ptr) complex_float_real_set
-@ stub -arch=arm ?real@?$_Complex_base@M@std@@QBAMXZ
+@ cdecl -arch=arm ?real@?$_Complex_base@M@std@@QBAMXZ(ptr) complex_float_real_get
 @ thiscall -arch=i386 ?real@?$_Complex_base@M@std@@QBEMXZ(ptr) complex_float_real_get
 @ cdecl -arch=win64 ?real@?$_Complex_base@M@std@@QEBAMXZ(ptr) complex_float_real_get
-@ stub -arch=arm ?real@?$_Complex_base@N@std@@QAANABN@Z
+@ cdecl -arch=arm ?real@?$_Complex_base@N@std@@QAANABN@Z(ptr ptr) complex_double_real_set
 @ thiscall -arch=i386 ?real@?$_Complex_base@N@std@@QAENABN@Z(ptr ptr) complex_double_real_set
 @ cdecl -arch=win64 ?real@?$_Complex_base@N@std@@QEAANAEBN@Z(ptr ptr) complex_double_real_set
-@ stub -arch=arm ?real@?$_Complex_base@N@std@@QBANXZ
+@ cdecl -arch=arm ?real@?$_Complex_base@N@std@@QBANXZ(ptr) complex_double_real_get
 @ thiscall -arch=i386 ?real@?$_Complex_base@N@std@@QBENXZ(ptr) complex_double_real_get
 @ cdecl -arch=win64 ?real@?$_Complex_base@N@std@@QEBANXZ(ptr) complex_double_real_get
-@ stub -arch=arm ?real@?$_Complex_base@O@std@@QAAOABO@Z
+@ cdecl -arch=arm ?real@?$_Complex_base@O@std@@QAAOABO@Z(ptr ptr) complex_double_real_set
 @ thiscall -arch=i386 ?real@?$_Complex_base@O@std@@QAEOABO@Z(ptr ptr) complex_double_real_set
 @ cdecl -arch=win64 ?real@?$_Complex_base@O@std@@QEAAOAEBO@Z(ptr ptr) complex_double_real_set
-@ stub -arch=arm ?real@?$_Complex_base@O@std@@QBAOXZ
+@ cdecl -arch=arm ?real@?$_Complex_base@O@std@@QBAOXZ(ptr) complex_double_real_get
 @ thiscall -arch=i386 ?real@?$_Complex_base@O@std@@QBEOXZ(ptr) complex_double_real_get
 @ cdecl -arch=win64 ?real@?$_Complex_base@O@std@@QEBAOXZ(ptr) complex_double_real_get
 @ cdecl -arch=win32 ?real@std@@YAMABV?$complex@M@1@@Z(ptr) complex_float_real
diff --git a/dlls/msvcp71/msvcp71.spec b/dlls/msvcp71/msvcp71.spec
index c7f4d60..55829c0 100644
--- a/dlls/msvcp71/msvcp71.spec
+++ b/dlls/msvcp71/msvcp71.spec
@@ -722,9 +722,9 @@
 @ cdecl -arch=arm ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAA@PAV?$basic_streambuf@_WU?$char_traits@_W@std@@@1@_N@Z(ptr ptr long long) basic_ostream_wchar_ctor
 @ thiscall -arch=i386 ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAE@PAV?$basic_streambuf@_WU?$char_traits@_W@std@@@1@_N@Z(ptr ptr long long) basic_ostream_wchar_ctor
 @ cdecl -arch=win64 ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QEAA@PEAV?$basic_streambuf@_WU?$char_traits@_W@std@@@1@_N@Z(ptr ptr long long) basic_ostream_wchar_ctor
-@ stub -arch=arm ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAA@W4_Uninitialized@1@@Z
-@ stub -arch=i386 ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAE@W4_Uninitialized@1@@Z
-@ stub -arch=win64 ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QEAA@W4_Uninitialized@1@@Z
+@ cdecl -arch=arm ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAA@W4_Uninitialized@1@@Z(ptr long long) basic_ostream_wchar_ctor_uninitialized_old
+@ thiscall -arch=i386 ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAE@W4_Uninitialized@1@@Z(ptr long long) basic_ostream_wchar_ctor_uninitialized_old
+@ cdecl -arch=win64 ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QEAA@W4_Uninitialized@1@@Z(ptr long long) basic_ostream_wchar_ctor_uninitialized_old
 @ cdecl -arch=arm ??0?$basic_ostringstream@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@QAA@ABV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@H@Z(ptr ptr long long) basic_ostringstream_char_ctor_str
 @ thiscall -arch=i386 ??0?$basic_ostringstream@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@QAE@ABV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@H@Z(ptr ptr long long) basic_ostringstream_char_ctor_str
 @ cdecl -arch=win64 ??0?$basic_ostringstream@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@QEAA@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@H@Z(ptr ptr long long) basic_ostringstream_char_ctor_str
@@ -1126,21 +1126,21 @@
 @ cdecl -arch=arm ??0?$num_put@_WV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@QAA@I@Z(ptr long) num_put_wchar_ctor_refs
 @ thiscall -arch=i386 ??0?$num_put@_WV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@QAE@I@Z(ptr long) num_put_wchar_ctor_refs
 @ cdecl -arch=win64 ??0?$num_put@_WV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@QEAA@_K@Z(ptr long) num_put_wchar_ctor_refs
-@ stub -arch=arm ??0?$numpunct@D@std@@QAA@ABV_Locinfo@1@I@Z
-@ stub -arch=i386 ??0?$numpunct@D@std@@QAE@ABV_Locinfo@1@I@Z
-@ stub -arch=win64 ??0?$numpunct@D@std@@QEAA@AEBV_Locinfo@1@_K@Z
+@ cdecl -arch=arm ??0?$numpunct@D@std@@QAA@ABV_Locinfo@1@I@Z(ptr ptr long) numpunct_char_ctor_locinfo_old
+@ thiscall -arch=i386 ??0?$numpunct@D@std@@QAE@ABV_Locinfo@1@I@Z(ptr ptr long) numpunct_char_ctor_locinfo_old
+@ cdecl -arch=win64 ??0?$numpunct@D@std@@QEAA@AEBV_Locinfo@1@_K@Z(ptr ptr long) numpunct_char_ctor_locinfo_old
 @ cdecl -arch=arm ??0?$numpunct@D@std@@QAA@I@Z(ptr long) numpunct_char_ctor_refs
 @ thiscall -arch=i386 ??0?$numpunct@D@std@@QAE@I@Z(ptr long) numpunct_char_ctor_refs
 @ cdecl -arch=win64 ??0?$numpunct@D@std@@QEAA@_K@Z(ptr long) numpunct_char_ctor_refs
-@ stub -arch=arm ??0?$numpunct@G@std@@QAA@ABV_Locinfo@1@I@Z
-@ stub -arch=i386 ??0?$numpunct@G@std@@QAE@ABV_Locinfo@1@I@Z
-@ stub -arch=win64 ??0?$numpunct@G@std@@QEAA@AEBV_Locinfo@1@_K@Z
+@ cdecl -arch=arm ??0?$numpunct@G@std@@QAA@ABV_Locinfo@1@I@Z(ptr ptr long) numpunct_short_ctor_locinfo_old
+@ thiscall -arch=i386 ??0?$numpunct@G@std@@QAE@ABV_Locinfo@1@I@Z(ptr ptr long) numpunct_short_ctor_locinfo_old
+@ cdecl -arch=win64 ??0?$numpunct@G@std@@QEAA@AEBV_Locinfo@1@_K@Z(ptr ptr long) numpunct_short_ctor_locinfo_old
 @ cdecl -arch=arm ??0?$numpunct@G@std@@QAA@I@Z(ptr long) numpunct_short_ctor_refs
 @ thiscall -arch=i386 ??0?$numpunct@G@std@@QAE@I@Z(ptr long) numpunct_short_ctor_refs
 @ cdecl -arch=win64 ??0?$numpunct@G@std@@QEAA@_K@Z(ptr long) numpunct_short_ctor_refs
-@ stub -arch=arm ??0?$numpunct@_W@std@@QAA@ABV_Locinfo@1@I@Z
-@ stub -arch=i386 ??0?$numpunct@_W@std@@QAE@ABV_Locinfo@1@I@Z
-@ stub -arch=win64 ??0?$numpunct@_W@std@@QEAA@AEBV_Locinfo@1@_K@Z
+@ cdecl -arch=arm ??0?$numpunct@_W@std@@QAA@ABV_Locinfo@1@I@Z(ptr ptr long) numpunct_wchar_ctor_locinfo_old
+@ thiscall -arch=i386 ??0?$numpunct@_W@std@@QAE@ABV_Locinfo@1@I@Z(ptr ptr long) numpunct_wchar_ctor_locinfo_old
+@ cdecl -arch=win64 ??0?$numpunct@_W@std@@QEAA@AEBV_Locinfo@1@_K@Z(ptr ptr long) numpunct_wchar_ctor_locinfo_old
 @ cdecl -arch=arm ??0?$numpunct@_W@std@@QAA@I@Z(ptr long) numpunct_wchar_ctor_refs
 @ thiscall -arch=i386 ??0?$numpunct@_W@std@@QAE@I@Z(ptr long) numpunct_wchar_ctor_refs
 @ cdecl -arch=win64 ??0?$numpunct@_W@std@@QEAA@_K@Z(ptr long) numpunct_wchar_ctor_refs
@@ -1797,7 +1797,7 @@
 @ cdecl -arch=arm ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAAAAV01@AAF@Z(ptr ptr) basic_istream_short_read_short
 @ thiscall -arch=i386 ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAEAAV01@AAF@Z(ptr ptr) basic_istream_short_read_short
 @ cdecl -arch=win64 ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QEAAAEAV01@AEAF@Z(ptr ptr) basic_istream_short_read_short
-@ stub -arch=arm ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAAAAV01@AAG@Z
+@ cdecl -arch=arm ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAAAAV01@AAG@Z(ptr ptr) basic_istream_wchar_read_ushort
 @ thiscall -arch=i386 ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAEAAV01@AAG@Z(ptr ptr) basic_istream_wchar_read_ushort
 @ cdecl -arch=win64 ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QEAAAEAV01@AEAG@Z(ptr ptr) basic_istream_wchar_read_ushort
 @ cdecl -arch=arm ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAAAAV01@AAH@Z(ptr ptr) basic_istream_short_read_int
@@ -2747,12 +2747,12 @@
 @ cdecl -arch=win64 ?_Getcat@?$numpunct@G@std@@SA_KPEAPEBVfacet@locale@2@@Z(ptr) numpunct_short__Getcat_old
 @ cdecl -arch=win32 ?_Getcat@?$numpunct@_W@std@@SAIPAPBVfacet@locale@2@@Z(ptr) numpunct_wchar__Getcat_old
 @ cdecl -arch=win64 ?_Getcat@?$numpunct@_W@std@@SA_KPEAPEBVfacet@locale@2@@Z(ptr) numpunct_wchar__Getcat_old
-@ stub -arch=win32 ?_Getcat@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@SAIPAPBVfacet@locale@2@@Z
-@ stub -arch=win64 ?_Getcat@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@SA_KPEAPEBVfacet@locale@2@@Z
-@ stub -arch=win32 ?_Getcat@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@SAIPAPBVfacet@locale@2@@Z
-@ stub -arch=win64 ?_Getcat@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@SA_KPEAPEBVfacet@locale@2@@Z
-@ stub -arch=win32 ?_Getcat@?$time_get@_WV?$istreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@SAIPAPBVfacet@locale@2@@Z
-@ stub -arch=win64 ?_Getcat@?$time_get@_WV?$istreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@SA_KPEAPEBVfacet@locale@2@@Z
+@ cdecl -arch=win32 ?_Getcat@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@SAIPAPBVfacet@locale@2@@Z(ptr) time_get_char__Getcat_old
+@ cdecl -arch=win64 ?_Getcat@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@SA_KPEAPEBVfacet@locale@2@@Z(ptr) time_get_char__Getcat_old
+@ cdecl -arch=win32 ?_Getcat@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@SAIPAPBVfacet@locale@2@@Z(ptr) time_get_short__Getcat_old
+@ cdecl -arch=win64 ?_Getcat@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@SA_KPEAPEBVfacet@locale@2@@Z(ptr) time_get_short__Getcat_old
+@ cdecl -arch=win32 ?_Getcat@?$time_get@_WV?$istreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@SAIPAPBVfacet@locale@2@@Z(ptr) time_get_wchar__Getcat_old
+@ cdecl -arch=win64 ?_Getcat@?$time_get@_WV?$istreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@SA_KPEAPEBVfacet@locale@2@@Z(ptr) time_get_wchar__Getcat_old
 @ cdecl -arch=win32 ?_Getcat@?$time_put@DV?$ostreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@SAIPAPBVfacet@locale@2@@Z(ptr) time_put_char__Getcat_old
 @ cdecl -arch=win64 ?_Getcat@?$time_put@DV?$ostreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@SA_KPEAPEBVfacet@locale@2@@Z(ptr) time_put_char__Getcat_old
 @ cdecl -arch=win32 ?_Getcat@?$time_put@GV?$ostreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@SAIPAPBVfacet@locale@2@@Z(ptr) time_put_short__Getcat_old
@@ -2994,15 +2994,15 @@
 @ cdecl -arch=arm ?_Init@?$num_put@_WV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@IAAXABV_Locinfo@2@@Z(ptr ptr) num_put_wchar__Init
 @ thiscall -arch=i386 ?_Init@?$num_put@_WV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@IAEXABV_Locinfo@2@@Z(ptr ptr) num_put_wchar__Init
 @ cdecl -arch=win64 ?_Init@?$num_put@_WV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@IEAAXAEBV_Locinfo@2@@Z(ptr ptr) num_put_wchar__Init
-@ stub -arch=arm ?_Init@?$numpunct@D@std@@IAAXABV_Locinfo@2@@Z
-@ stub -arch=i386 ?_Init@?$numpunct@D@std@@IAEXABV_Locinfo@2@@Z
-@ stub -arch=win64 ?_Init@?$numpunct@D@std@@IEAAXAEBV_Locinfo@2@@Z
-@ stub -arch=arm ?_Init@?$numpunct@G@std@@IAAXABV_Locinfo@2@@Z
-@ stub -arch=i386 ?_Init@?$numpunct@G@std@@IAEXABV_Locinfo@2@@Z
-@ stub -arch=win64 ?_Init@?$numpunct@G@std@@IEAAXAEBV_Locinfo@2@@Z
-@ stub -arch=arm ?_Init@?$numpunct@_W@std@@IAAXABV_Locinfo@2@@Z
-@ stub -arch=i386 ?_Init@?$numpunct@_W@std@@IAEXABV_Locinfo@2@@Z
-@ stub -arch=win64 ?_Init@?$numpunct@_W@std@@IEAAXAEBV_Locinfo@2@@Z
+@ cdecl -arch=arm ?_Init@?$numpunct@D@std@@IAAXABV_Locinfo@2@@Z(ptr ptr) numpunct_char__Init_old
+@ thiscall -arch=i386 ?_Init@?$numpunct@D@std@@IAEXABV_Locinfo@2@@Z(ptr ptr) numpunct_char__Init_old
+@ cdecl -arch=win64 ?_Init@?$numpunct@D@std@@IEAAXAEBV_Locinfo@2@@Z(ptr ptr) numpunct_char__Init_old
+@ cdecl -arch=arm ?_Init@?$numpunct@G@std@@IAAXABV_Locinfo@2@@Z(ptr ptr) numpunct_wchar__Init_old
+@ thiscall -arch=i386 ?_Init@?$numpunct@G@std@@IAEXABV_Locinfo@2@@Z(ptr ptr) numpunct_wchar__Init_old
+@ cdecl -arch=win64 ?_Init@?$numpunct@G@std@@IEAAXAEBV_Locinfo@2@@Z(ptr ptr) numpunct_wchar__Init_old
+@ cdecl -arch=arm ?_Init@?$numpunct@_W@std@@IAAXABV_Locinfo@2@@Z(ptr ptr) numpunct_wchar__Init_old
+@ thiscall -arch=i386 ?_Init@?$numpunct@_W@std@@IAEXABV_Locinfo@2@@Z(ptr ptr) numpunct_wchar__Init_old
+@ cdecl -arch=win64 ?_Init@?$numpunct@_W@std@@IEAAXAEBV_Locinfo@2@@Z(ptr ptr) numpunct_wchar__Init_old
 @ cdecl -arch=arm ?_Init@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@IAAXABV_Locinfo@2@@Z(ptr ptr) time_get_char__Init
 @ thiscall -arch=i386 ?_Init@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@IAEXABV_Locinfo@2@@Z(ptr ptr) time_get_char__Init
 @ cdecl -arch=win64 ?_Init@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@IEAAXAEBV_Locinfo@2@@Z(ptr ptr) time_get_char__Init
diff --git a/dlls/msvcp80/msvcp80.spec b/dlls/msvcp80/msvcp80.spec
index 8c84632..199883d 100644
--- a/dlls/msvcp80/msvcp80.spec
+++ b/dlls/msvcp80/msvcp80.spec
@@ -818,9 +818,9 @@
 @ cdecl -arch=arm ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAA@PAV?$basic_streambuf@_WU?$char_traits@_W@std@@@1@_N@Z(ptr ptr long long) basic_ostream_wchar_ctor
 @ thiscall -arch=i386 ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAE@PAV?$basic_streambuf@_WU?$char_traits@_W@std@@@1@_N@Z(ptr ptr long long) basic_ostream_wchar_ctor
 @ cdecl -arch=win64 ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QEAA@PEAV?$basic_streambuf@_WU?$char_traits@_W@std@@@1@_N@Z(ptr ptr long long) basic_ostream_wchar_ctor
-@ stub -arch=arm ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAA@W4_Uninitialized@1@@Z
-@ stub -arch=i386 ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAE@W4_Uninitialized@1@@Z
-@ stub -arch=win64 ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QEAA@W4_Uninitialized@1@@Z
+@ cdecl -arch=arm ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAA@W4_Uninitialized@1@@Z(ptr long long) basic_ostream_wchar_ctor_uninitialized_old
+@ thiscall -arch=i386 ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAE@W4_Uninitialized@1@@Z(ptr long long) basic_ostream_wchar_ctor_uninitialized_old
+@ cdecl -arch=win64 ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QEAA@W4_Uninitialized@1@@Z(ptr long long) basic_ostream_wchar_ctor_uninitialized_old
 @ cdecl -arch=arm ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAA@W4_Uninitialized@1@_N@Z(ptr long long long) basic_ostream_wchar_ctor_uninitialized
 @ thiscall -arch=i386 ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAE@W4_Uninitialized@1@_N@Z(ptr long long long) basic_ostream_wchar_ctor_uninitialized
 @ cdecl -arch=win64 ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QEAA@W4_Uninitialized@1@_N@Z(ptr long long long) basic_ostream_wchar_ctor_uninitialized
@@ -1271,30 +1271,30 @@
 @ cdecl -arch=arm ??0?$num_put@_WV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@QAA@I@Z(ptr long) num_put_wchar_ctor_refs
 @ thiscall -arch=i386 ??0?$num_put@_WV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@QAE@I@Z(ptr long) num_put_wchar_ctor_refs
 @ cdecl -arch=win64 ??0?$num_put@_WV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@QEAA@_K@Z(ptr long) num_put_wchar_ctor_refs
-@ stub -arch=arm ??0?$numpunct@D@std@@IAA@PBDI@Z
-@ stub -arch=i386 ??0?$numpunct@D@std@@IAE@PBDI@Z
-@ stub -arch=win64 ??0?$numpunct@D@std@@IEAA@PEBD_K@Z
-@ stub -arch=arm ??0?$numpunct@D@std@@QAA@ABV_Locinfo@1@I@Z
-@ stub -arch=i386 ??0?$numpunct@D@std@@QAE@ABV_Locinfo@1@I@Z
-@ stub -arch=win64 ??0?$numpunct@D@std@@QEAA@AEBV_Locinfo@1@_K@Z
+@ cdecl -arch=arm ??0?$numpunct@D@std@@IAA@PBDI@Z(ptr str long) numpunct_char_ctor_name_old
+@ thiscall -arch=i386 ??0?$numpunct@D@std@@IAE@PBDI@Z(ptr str long) numpunct_char_ctor_name_old
+@ cdecl -arch=win64 ??0?$numpunct@D@std@@IEAA@PEBD_K@Z(ptr str long) numpunct_char_ctor_name_old
+@ cdecl -arch=arm ??0?$numpunct@D@std@@QAA@ABV_Locinfo@1@I@Z(ptr ptr long) numpunct_char_ctor_locinfo_old
+@ thiscall -arch=i386 ??0?$numpunct@D@std@@QAE@ABV_Locinfo@1@I@Z(ptr ptr long) numpunct_char_ctor_locinfo_old
+@ cdecl -arch=win64 ??0?$numpunct@D@std@@QEAA@AEBV_Locinfo@1@_K@Z(ptr ptr long) numpunct_char_ctor_locinfo_old
 @ cdecl -arch=arm ??0?$numpunct@D@std@@QAA@I@Z(ptr long) numpunct_char_ctor_refs
 @ thiscall -arch=i386 ??0?$numpunct@D@std@@QAE@I@Z(ptr long) numpunct_char_ctor_refs
 @ cdecl -arch=win64 ??0?$numpunct@D@std@@QEAA@_K@Z(ptr long) numpunct_char_ctor_refs
-@ stub -arch=arm ??0?$numpunct@G@std@@IAA@PBDI@Z
-@ stub -arch=i386 ??0?$numpunct@G@std@@IAE@PBDI@Z
-@ stub -arch=win64 ??0?$numpunct@G@std@@IEAA@PEBD_K@Z
-@ stub -arch=arm ??0?$numpunct@G@std@@QAA@ABV_Locinfo@1@I@Z
-@ stub -arch=i386 ??0?$numpunct@G@std@@QAE@ABV_Locinfo@1@I@Z
-@ stub -arch=win64 ??0?$numpunct@G@std@@QEAA@AEBV_Locinfo@1@_K@Z
+@ cdecl -arch=arm ??0?$numpunct@G@std@@IAA@PBDI@Z(ptr str long) numpunct_short_ctor_name_old
+@ thiscall -arch=i386 ??0?$numpunct@G@std@@IAE@PBDI@Z(ptr str long) numpunct_short_ctor_name_old
+@ cdecl -arch=win64 ??0?$numpunct@G@std@@IEAA@PEBD_K@Z(ptr str long) numpunct_short_ctor_name_old
+@ cdecl -arch=arm ??0?$numpunct@G@std@@QAA@ABV_Locinfo@1@I@Z(ptr ptr long) numpunct_short_ctor_locinfo_old
+@ thiscall -arch=i386 ??0?$numpunct@G@std@@QAE@ABV_Locinfo@1@I@Z(ptr ptr long) numpunct_short_ctor_locinfo_old
+@ cdecl -arch=win64 ??0?$numpunct@G@std@@QEAA@AEBV_Locinfo@1@_K@Z(ptr ptr long) numpunct_short_ctor_locinfo_old
 @ cdecl -arch=arm ??0?$numpunct@G@std@@QAA@I@Z(ptr long) numpunct_short_ctor_refs
 @ thiscall -arch=i386 ??0?$numpunct@G@std@@QAE@I@Z(ptr long) numpunct_short_ctor_refs
 @ cdecl -arch=win64 ??0?$numpunct@G@std@@QEAA@_K@Z(ptr long) numpunct_short_ctor_refs
-@ stub -arch=arm ??0?$numpunct@_W@std@@IAA@PBDI@Z
-@ stub -arch=i386 ??0?$numpunct@_W@std@@IAE@PBDI@Z
-@ stub -arch=win64 ??0?$numpunct@_W@std@@IEAA@PEBD_K@Z
-@ stub -arch=arm ??0?$numpunct@_W@std@@QAA@ABV_Locinfo@1@I@Z
-@ stub -arch=i386 ??0?$numpunct@_W@std@@QAE@ABV_Locinfo@1@I@Z
-@ stub -arch=win64 ??0?$numpunct@_W@std@@QEAA@AEBV_Locinfo@1@_K@Z
+@ cdecl -arch=arm ??0?$numpunct@_W@std@@IAA@PBDI@Z(ptr str long) numpunct_wchar_ctor_name_old
+@ thiscall -arch=i386 ??0?$numpunct@_W@std@@IAE@PBDI@Z(ptr str long) numpunct_wchar_ctor_name_old
+@ cdecl -arch=win64 ??0?$numpunct@_W@std@@IEAA@PEBD_K@Z(ptr str long) numpunct_wchar_ctor_name_old
+@ cdecl -arch=arm ??0?$numpunct@_W@std@@QAA@ABV_Locinfo@1@I@Z(ptr ptr long) numpunct_wchar_ctor_locinfo_old
+@ thiscall -arch=i386 ??0?$numpunct@_W@std@@QAE@ABV_Locinfo@1@I@Z(ptr ptr long) numpunct_wchar_ctor_locinfo_old
+@ cdecl -arch=win64 ??0?$numpunct@_W@std@@QEAA@AEBV_Locinfo@1@_K@Z(ptr ptr long) numpunct_wchar_ctor_locinfo_old
 @ cdecl -arch=arm ??0?$numpunct@_W@std@@QAA@I@Z(ptr long) numpunct_wchar_ctor_refs
 @ thiscall -arch=i386 ??0?$numpunct@_W@std@@QAE@I@Z(ptr long) numpunct_wchar_ctor_refs
 @ cdecl -arch=win64 ??0?$numpunct@_W@std@@QEAA@_K@Z(ptr long) numpunct_wchar_ctor_refs
@@ -2009,7 +2009,7 @@
 @ cdecl -arch=arm ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAAAAV01@AAF@Z(ptr ptr) basic_istream_short_read_short
 @ thiscall -arch=i386 ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAEAAV01@AAF@Z(ptr ptr) basic_istream_short_read_short
 @ cdecl -arch=win64 ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QEAAAEAV01@AEAF@Z(ptr ptr) basic_istream_short_read_short
-@ stub -arch=arm ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAAAAV01@AAG@Z
+@ cdecl -arch=arm ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAAAAV01@AAG@Z(ptr ptr) basic_istream_wchar_read_ushort
 @ thiscall -arch=i386 ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAEAAV01@AAG@Z(ptr ptr) basic_istream_wchar_read_ushort
 @ cdecl -arch=win64 ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QEAAAEAV01@AEAG@Z(ptr ptr) basic_istream_wchar_read_ushort
 @ cdecl -arch=arm ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAAAAV01@AAH@Z(ptr ptr) basic_istream_short_read_int
@@ -3052,12 +3052,12 @@
 @ cdecl -arch=win64 ?_Getcat@?$numpunct@G@std@@SA_KPEAPEBVfacet@locale@2@@Z(ptr) numpunct_short__Getcat_old
 @ cdecl -arch=win32 ?_Getcat@?$numpunct@_W@std@@SAIPAPBVfacet@locale@2@@Z(ptr) numpunct_wchar__Getcat_old
 @ cdecl -arch=win64 ?_Getcat@?$numpunct@_W@std@@SA_KPEAPEBVfacet@locale@2@@Z(ptr) numpunct_wchar__Getcat_old
-@ stub -arch=win32 ?_Getcat@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@SAIPAPBVfacet@locale@2@@Z
-@ stub -arch=win64 ?_Getcat@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@SA_KPEAPEBVfacet@locale@2@@Z
-@ stub -arch=win32 ?_Getcat@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@SAIPAPBVfacet@locale@2@@Z
-@ stub -arch=win64 ?_Getcat@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@SA_KPEAPEBVfacet@locale@2@@Z
-@ stub -arch=win32 ?_Getcat@?$time_get@_WV?$istreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@SAIPAPBVfacet@locale@2@@Z
-@ stub -arch=win64 ?_Getcat@?$time_get@_WV?$istreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@SA_KPEAPEBVfacet@locale@2@@Z
+@ cdecl -arch=win32 ?_Getcat@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@SAIPAPBVfacet@locale@2@@Z(ptr) time_get_char__Getcat_old
+@ cdecl -arch=win64 ?_Getcat@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@SA_KPEAPEBVfacet@locale@2@@Z(ptr) time_get_char__Getcat_old
+@ cdecl -arch=win32 ?_Getcat@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@SAIPAPBVfacet@locale@2@@Z(ptr) time_get_short__Getcat_old
+@ cdecl -arch=win64 ?_Getcat@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@SA_KPEAPEBVfacet@locale@2@@Z(ptr) time_get_short__Getcat_old
+@ cdecl -arch=win32 ?_Getcat@?$time_get@_WV?$istreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@SAIPAPBVfacet@locale@2@@Z(ptr) time_get_wchar__Getcat_old
+@ cdecl -arch=win64 ?_Getcat@?$time_get@_WV?$istreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@SA_KPEAPEBVfacet@locale@2@@Z(ptr) time_get_wchar__Getcat_old
 @ cdecl -arch=win32 ?_Getcat@?$time_put@DV?$ostreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@SAIPAPBVfacet@locale@2@@Z(ptr) time_put_char__Getcat_old
 @ cdecl -arch=win64 ?_Getcat@?$time_put@DV?$ostreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@SA_KPEAPEBVfacet@locale@2@@Z(ptr) time_put_char__Getcat_old
 @ cdecl -arch=win32 ?_Getcat@?$time_put@GV?$ostreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@SAIPAPBVfacet@locale@2@@Z(ptr) time_put_short__Getcat_old
@@ -3333,15 +3333,15 @@
 @ cdecl -arch=arm ?_Init@?$num_put@_WV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@IAAXABV_Locinfo@2@@Z(ptr ptr) num_put_wchar__Init
 @ thiscall -arch=i386 ?_Init@?$num_put@_WV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@IAEXABV_Locinfo@2@@Z(ptr ptr) num_put_wchar__Init
 @ cdecl -arch=win64 ?_Init@?$num_put@_WV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@IEAAXAEBV_Locinfo@2@@Z(ptr ptr) num_put_wchar__Init
-@ stub -arch=arm ?_Init@?$numpunct@D@std@@IAAXABV_Locinfo@2@@Z
-@ stub -arch=i386 ?_Init@?$numpunct@D@std@@IAEXABV_Locinfo@2@@Z
-@ stub -arch=win64 ?_Init@?$numpunct@D@std@@IEAAXAEBV_Locinfo@2@@Z
-@ stub -arch=arm ?_Init@?$numpunct@G@std@@IAAXABV_Locinfo@2@@Z
-@ stub -arch=i386 ?_Init@?$numpunct@G@std@@IAEXABV_Locinfo@2@@Z
-@ stub -arch=win64 ?_Init@?$numpunct@G@std@@IEAAXAEBV_Locinfo@2@@Z
-@ stub -arch=arm ?_Init@?$numpunct@_W@std@@IAAXABV_Locinfo@2@@Z
-@ stub -arch=i386 ?_Init@?$numpunct@_W@std@@IAEXABV_Locinfo@2@@Z
-@ stub -arch=win64 ?_Init@?$numpunct@_W@std@@IEAAXAEBV_Locinfo@2@@Z
+@ cdecl -arch=arm ?_Init@?$numpunct@D@std@@IAAXABV_Locinfo@2@@Z(ptr ptr) numpunct_char__Init_old
+@ thiscall -arch=i386 ?_Init@?$numpunct@D@std@@IAEXABV_Locinfo@2@@Z(ptr ptr) numpunct_char__Init_old
+@ cdecl -arch=win64 ?_Init@?$numpunct@D@std@@IEAAXAEBV_Locinfo@2@@Z(ptr ptr) numpunct_char__Init_old
+@ cdecl -arch=arm ?_Init@?$numpunct@G@std@@IAAXABV_Locinfo@2@@Z(ptr ptr) numpunct_wchar__Init_old
+@ thiscall -arch=i386 ?_Init@?$numpunct@G@std@@IAEXABV_Locinfo@2@@Z(ptr ptr) numpunct_wchar__Init_old
+@ cdecl -arch=win64 ?_Init@?$numpunct@G@std@@IEAAXAEBV_Locinfo@2@@Z(ptr ptr) numpunct_wchar__Init_old
+@ cdecl -arch=arm ?_Init@?$numpunct@_W@std@@IAAXABV_Locinfo@2@@Z(ptr ptr) numpunct_wchar__Init_old
+@ thiscall -arch=i386 ?_Init@?$numpunct@_W@std@@IAEXABV_Locinfo@2@@Z(ptr ptr) numpunct_wchar__Init_old
+@ cdecl -arch=win64 ?_Init@?$numpunct@_W@std@@IEAAXAEBV_Locinfo@2@@Z(ptr ptr) numpunct_wchar__Init_old
 @ cdecl -arch=arm ?_Init@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@IAAXABV_Locinfo@2@@Z(ptr ptr) time_get_char__Init
 @ thiscall -arch=i386 ?_Init@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@IAEXABV_Locinfo@2@@Z(ptr ptr) time_get_char__Init
 @ cdecl -arch=win64 ?_Init@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@IEAAXAEBV_Locinfo@2@@Z(ptr ptr) time_get_char__Init
diff --git a/dlls/msvcp90/ios.c b/dlls/msvcp90/ios.c
index 879772b..28bc307 100644
--- a/dlls/msvcp90/ios.c
+++ b/dlls/msvcp90/ios.c
@@ -7351,6 +7351,18 @@ basic_ostream_wchar* __thiscall basic_ostream_wchar_ctor_uninitialized(basic_ost
     return this;
 }
 
+#if _MSVCP_VER >= 70 && _MSVCP_VER <= 80
+/* ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QAE@W4_Uninitialized@1@@Z */
+/* ??0?$basic_ostream@_WU?$char_traits@_W@std@@@std@@QEAA@W4_Uninitialized@1@@Z */
+/* the older form, without the flag: the stream is always one of the standard ones */
+DEFINE_THISCALL_WRAPPER(basic_ostream_wchar_ctor_uninitialized_old, 12)
+basic_ostream_wchar* __thiscall basic_ostream_wchar_ctor_uninitialized_old(basic_ostream_wchar *this,
+        int uninitialized, bool virt_init)
+{
+    return basic_ostream_wchar_ctor_uninitialized(this, uninitialized, TRUE, virt_init);
+}
+#endif
+
 /* ??0?$basic_ostream@GU?$char_traits@G@std@@@std@@QAE@W4_Uninitialized@1@_N@Z */
 /* ??0?$basic_ostream@GU?$char_traits@G@std@@@std@@QEAA@W4_Uninitialized@1@_N@Z */
 DEFINE_THISCALL_WRAPPER(basic_ostream_short_ctor_uninitialized, 16)
diff --git a/dlls/msvcp90/locale.c b/dlls/msvcp90/locale.c
index 7381306..2682614 100644
--- a/dlls/msvcp90/locale.c
+++ b/dlls/msvcp90/locale.c
@@ -16815,3 +16815,94 @@ void free_locale(void)
         operator_delete(iter);
     }
 }
+
+#if _MSVCP_VER <= 80
+/* The forms of the numpunct constructors and _Init without the "use the
+ * default names" flag, and of time_get::_Getcat without a locale, that the
+ * older libraries export. */
+
+/* ??0?$numpunct@D@std@@QAE@ABV_Locinfo@1@I@Z */
+/* ??0?$numpunct@D@std@@QEAA@AEBV_Locinfo@1@_K@Z */
+DEFINE_THISCALL_WRAPPER(numpunct_char_ctor_locinfo_old, 12)
+numpunct_char* __thiscall numpunct_char_ctor_locinfo_old(numpunct_char *this, const _Locinfo *locinfo, size_t refs)
+{
+    return numpunct_char_ctor_locinfo(this, locinfo, refs, FALSE);
+}
+
+/* ??0?$numpunct@_W@std@@QAE@ABV_Locinfo@1@I@Z */
+/* ??0?$numpunct@_W@std@@QEAA@AEBV_Locinfo@1@_K@Z */
+DEFINE_THISCALL_WRAPPER(numpunct_wchar_ctor_locinfo_old, 12)
+numpunct_wchar* __thiscall numpunct_wchar_ctor_locinfo_old(numpunct_wchar *this, const _Locinfo *locinfo, size_t refs)
+{
+    return numpunct_wchar_ctor_locinfo(this, locinfo, refs, FALSE);
+}
+
+/* ??0?$numpunct@G@std@@QAE@ABV_Locinfo@1@I@Z */
+/* ??0?$numpunct@G@std@@QEAA@AEBV_Locinfo@1@_K@Z */
+DEFINE_THISCALL_WRAPPER(numpunct_short_ctor_locinfo_old, 12)
+numpunct_wchar* __thiscall numpunct_short_ctor_locinfo_old(numpunct_wchar *this, const _Locinfo *locinfo, size_t refs)
+{
+    return numpunct_short_ctor_locinfo(this, locinfo, refs, FALSE);
+}
+
+/* ??0?$numpunct@D@std@@IAE@PBDI@Z */
+/* ??0?$numpunct@D@std@@IEAA@PEBD_K@Z */
+DEFINE_THISCALL_WRAPPER(numpunct_char_ctor_name_old, 12)
+numpunct_char* __thiscall numpunct_char_ctor_name_old(numpunct_char *this, const char *name, size_t refs)
+{
+    return numpunct_char_ctor_name(this, name, refs, FALSE);
+}
+
+/* ??0?$numpunct@_W@std@@IAE@PBDI@Z */
+/* ??0?$numpunct@_W@std@@IEAA@PEBD_K@Z */
+DEFINE_THISCALL_WRAPPER(numpunct_wchar_ctor_name_old, 12)
+numpunct_wchar* __thiscall numpunct_wchar_ctor_name_old(numpunct_wchar *this, const char *name, size_t refs)
+{
+    return numpunct_wchar_ctor_name(this, name, refs, FALSE);
+}
+
+/* ??0?$numpunct@G@std@@IAE@PBDI@Z */
+/* ??0?$numpunct@G@std@@IEAA@PEBD_K@Z */
+DEFINE_THISCALL_WRAPPER(numpunct_short_ctor_name_old, 12)
+numpunct_wchar* __thiscall numpunct_short_ctor_name_old(numpunct_wchar *this, const char *name, size_t refs)
+{
+    return numpunct_short_ctor_name(this, name, refs, FALSE);
+}
+
+/* ?_Init@?$numpunct@D@std@@IAEXABV_Locinfo@2@@Z */
+/* ?_Init@?$numpunct@D@std@@IEAAXAEBV_Locinfo@2@@Z */
+DEFINE_THISCALL_WRAPPER(numpunct_char__Init_old, 8)
+void __thiscall numpunct_char__Init_old(numpunct_char *this, const _Locinfo *locinfo)
+{
+    numpunct_char__Init(this, locinfo, FALSE);
+}
+
+/* ?_Init@?$numpunct@_W@std@@IAEXABV_Locinfo@2@@Z */
+/* ?_Init@?$numpunct@G@std@@IAEXABV_Locinfo@2@@Z */
+DEFINE_THISCALL_WRAPPER(numpunct_wchar__Init_old, 8)
+void __thiscall numpunct_wchar__Init_old(numpunct_wchar *this, const _Locinfo *locinfo)
+{
+    numpunct_wchar__Init(this, locinfo, FALSE);
+}
+
+/* ?_Getcat@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@SAIPAPBVfacet@locale@2@@Z */
+unsigned int __cdecl time_get_char__Getcat_old(const locale_facet **facet)
+{
+#ifdef SG_MUTANT_TIME_GET_GETCAT_OLD_NONE
+    return LC_TIME;
+#endif
+    return time_get_char__Getcat(facet, locale_classic());
+}
+
+/* ?_Getcat@?$time_get@_WV?$istreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@SAIPAPBVfacet@locale@2@@Z */
+unsigned int __cdecl time_get_wchar__Getcat_old(const locale_facet **facet)
+{
+    return time_get_wchar__Getcat(facet, locale_classic());
+}
+
+/* ?_Getcat@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@SAIPAPBVfacet@locale@2@@Z */
+unsigned int __cdecl time_get_short__Getcat_old(const locale_facet **facet)
+{
+    return time_get_short__Getcat(facet, locale_classic());
+}
+#endif
